### ISSUE TITLE
err:vkd3d_create_vk_device: Failed to create Vulkan device, vr -3 on RTX 3090

### ISSUE THREAD

**Issue by RoyShapiro** [HansKristian-Work/vkd3d-proton#488]
Hi!
"First world" problems here.
After the D3D12 Ultimate FL update (specifically vkd3d-proton built from commit https://github.com/HansKristian-Work/vkd3d-proton/commit/b4bc92714a1f2bbd3bef69b2e0c07ac1b19f54a7), I get error:
`160:err:vkd3d_create_vk_device: Failed to create Vulkan device, vr -3.`
On all D3D12 games (tried Cyberpunk, Serious Sam 4, Resident Evils), and they all fail to start. First number, 160, is different for each game and can be 244, e.t.c.
Running the following:
Software:
Debian 10 Buster, Kernel 5.10.11-xanmod1 (tried 5.9, same story)
NVidia Driver 455.50.04 (started with 455.46.02, same story)
Vulkan Loader 1.2.166 (started with 1.2.162, same story)
Proton 5.13 or current Proton Experimental (neither works).
Hardware:
AMD Ryzen Threadripper 1920X,
64Gb RAM 3000 Mhz,
Inno3D iChill X3 GeForce RTX 3090.
All games I've tried work fine if I rollback to Vkd3d-Proton v.2.1 official release package dll. The problem starts with this newest commit build.
Not posting any specific logs yet, because it affects all D3D12 games equally, and the rollback also works for all of them.
Any advice?

**Comment by doitsujin** [comment]
Can't reproduce on Turing (RTX 2060) or any AMD hardware.

This is the second time someone reports this on an Ampere GPU though. Apparently the Nvidia driver just refuses to work on those cards when we enable the raytracing extensions.

**Comment by RoyShapiro** [comment]
Yes, that's what I suspected as well. Question is, why. Strange that both
driver versions don't work, Khronos Vulkan Extensions are advertised as
enabled since version 455.46.02. Maybe there's some specific Ampere-related
quirk in the driver that can be worked around somehow.

**Comment by doitsujin** [comment]
Can you test if https://github.com/HansKristian-Work/vkd3d-proton/pull/489 fixes the problem?

**Comment by RoyShapiro** [comment]
Probably. Give me half an hour, I'll return with the results & update this
comment.

**Comment by RoyShapiro** [comment]
@doitsujin Alas, no dice. Same exact error.

**Comment by liam-middlebrook** [comment]
Roy, could you run `/usr/bin/nvidia-bug-report.sh` and either post the log.gz file it generates here or in an email to `linux-bugs [at] nvidia.com` (and mention this issue in the email)?

**Comment by RoyShapiro** [comment]
@liam-middlebrook Sure. Here. 
[nvidia-bug-report.log.gz](https://github.com/HansKristian-Work/vkd3d-proton/files/5889207/nvidia-bug-report.log.gz)

**Comment by doitsujin** [comment]
Can you test if the following works:
```
meson -Denable_tests=true build
cd build
ninja
./tests/d3d12
```
This builds vkd3d-proton and our test suite as a native Linux binary, thus bypassing winevulkan. Would be strange if that was the issue, but if it spews a lot of messages about failing to create a device as well, we at least know for sure it's not that.

**Comment by RoyShapiro** [comment]
Yep. Did it. Got segfault running said test. Full output to console attached.
[output.txt](https://github.com/HansKristian-Work/vkd3d-proton/files/5894934/output.txt)
_Since I usually compile the project as a Windows DLL (and hence using MinGW64 on a virtual machine), and this time around I built it on bare metal (so it built natively), I could've missed something important. For instance, I did not explicitly install or built glslang compiler. However, there were no build errors, so it is either already present (installed as a dependency by something or another) or it didn't explicitly tell me, that I'm doing something dumb. If this is the issue, let me know, I'll see what I can do._

**Comment by doitsujin** [comment]
It succeeds at creating a device though, so that's... interesting.

**Comment by liam-middlebrook** [comment]
The above datapoint for a native Linux vkd3d-proton matches my findings so far (minus the bit about the segfault). This issue isn't vkd3d-proton or Proton specific, I was able to reproduce it with Quake II RTX, both when running through Proton as well as the Steam Linux Runtime (more details below).


The difference here that is specific to NVIDIA Ampere GPUs as opposed to older GPU generations is the backend the driver will use as part of Ray Tracing operations. In the failing case this newer backend isn't being loaded properly. From my experimentation it seems that this issue is isolated to the Steam Linux Runtime, which Proton 5.13 and above are built on top of.

Since it seems like this may be blocking you from enjoying your games, one thing you could do is to use a custom Proton build (like a TK-Glitche variant) which doesn't make use of the Steam Linux Runtime, or I believe there are some shell script edits that people have done in the past back when the new runtime had issues with Vulkan layers like MangoHUD.

I've been in contact with some folks who work on the Steam Linux Runtime to better understand the issue here. It's still unknown whether the bug is in the driver or the runtime, but from my understanding this issue should not effect applications which are not making use of the runtime.

**Comment by RoyShapiro** [comment]
Okay, Liam, I admit that I did not yet try it with pure Wine, the games I
tested do make use of either Stream (Resident Evils, et al) or Lutris
(Serious Sam 4 @ GOG) Runtime. I will try to run one or the other without
the Runtime and return with the results.

**Comment by RoyShapiro** [comment]
@doitsujin @liam-middlebrook Liam, you were correct, when run without the Steam \ Lutris runtime, it works fine. 

_DXR is still disabled, I see. I will admit, after Liam's advice worked, I tried enabling it on my own risk by editing device.c, it shows in the menu, but just crashes the game afterwards (tried it on Cyberpunk by setting reported tier to TIER_1_0 - sorry, my bad, got impatient), so I've re-disabled it for now._

**Comment by liam-middlebrook** [comment]
NVIDIA Vulkan Developer Beta driver release [455.50.05](https://developer.nvidia.com/vulkan-driver) contains a fix for this issue.

**Comment by RoyShapiro** [comment]
@liam-middlebrook Thanks.
To anyone reading this issue, it's important to note, that in order for this to work, it is necessary to not only update the drivers but to also rebuild vkd3d-proton using the latest commits. For me https://github.com/HansKristian-Work/vkd3d-proton/commit/51bf9397433fa0d002ff65b4bea327a7cb7d790f worked.

**Comment by RoyShapiro** [comment]
@doitsujin @liam-middlebrook 
Hello again.
As it often happens further testing **revealed more problems**.
With the driver 455.50.05 update the issue should've been fixed, but as it turns out, _not quite_.
The behavior I get now is very _peculiar_ and _interesting_:

After my initial testing, which - incidently - I've done after running ./tests/d3d12 as mentioned in https://github.com/HansKristian-Work/vkd3d-proton/issues/488#issuecomment-769799156 - an _important detail_ further on, I have rebooted my system for unrelated reasons.

When I've tried launching D3D12 apps after said reboot, I was again, to my utter surprise, greeted with
`err:vkd3d_create_vk_device: Failed to create Vulkan device, vr -3.`
on _all_ of them, that use any post-DXR commit vkd3d-proton build.

Seeing this I've tried running said apps without Steam Runtime, which helped previously. _This time it DID NOT help._
Apps refused to run with `err:vkd3d_create_vk_device: Failed to create Vulkan device, vr -3.` error **with or without** Steam runtime with 455.50.05 drivers.

But they did work the first time right after installing the new driver with or without Steam Runtime. So, I began retracing my steps afterwards to try and find the reason why they worked, then suddenly stopped.
Long story short, I've remembered, that for whatever reason I did re-run `./tests/d3d12` from vkd3d-proton almost right after installing the new drivers. So, crazy as the idea was, I've tried running it again. It's output seemingly didn't change, and as mentioned above, it ran for some time, then crashed with `Segmentation fault` error.
However, **right after** that I've immediately tried running some D3D12 apps. And... They _all_ worked! _With or without_ the Steam Runtime.

Perplexed, I've tried rebooting again. And just as I've expected, _nothing_ worked again, and yet again with the dreaded
`err:vkd3d_create_vk_device: Failed to create Vulkan device, vr -3.` error. So, I've tried running `./tests/d3d12` again, and once again, as soon as it finished working, everything worked like a charm.
I've tried this same algorithm several times, and can confirm that:

- On a freshly rebooted system with driver 455.50.05:
Nothing works, error `err:vkd3d_create_vk_device: Failed to create Vulkan device, vr -3.`

- However if, _at any point in time after reboot_, I run `./tests/d3d12` from vkd3d-proton tests suite, everything starts working afterwards and keeps working until the next reboot. This helps (and happens) _every single time_.

This is a very unexpected turn of events. It would seem, as if that there's some problem going on with
`winevulkan`
that happens on every attempt to launch it when the system is freshly rebooted.
However, running the `native` implementation of vkd3d-proton from `./tests/d3d12` sort of "_preheats_", for the lack of a better term, that subsystem for launch, likely by changing certain variables on the GPU side (somehow), that result in everything working afterwards.

And yes, I did try running other native Vulkan apps instead. But all of them launch successfully without any issues, and none of them (at least those I've tried) make `err:vkd3d_create_vk_device: Failed to create Vulkan device, vr -3.` go away, except for `./tests/d3d12`.

Fresh console output from `./tests/d3d12` is attached (it is after this output that everything starts working):
[output.txt](https://github.com/HansKristian-Work/vkd3d-proton/files/5934951/output.txt)

I think it might be worth reopening this issue, as it doesn't seem clear, where the actual source of such an unusual behavior lies.
